Mathlib/SetTheory/Game/Nim.lean

```diff
@@ -313,68 +317,56 @@ theorem grundyValue_ne_moveRight {G : PGame} [G.Impartial] (i : G.RightMoves) :
-      obtain ⟨k, hk⟩ := lt_omega0.1 (hj.trans (nat_lt_omega0 _))
-      rw [hk, Nat.cast_lt] at hj
-      have := hj.ne
-      have := hj -- The termination checker doesn't work without this.
-    · rwa [add_moveLeft_inl, moveLeft_nim', ne_eq, hk, grundyValue_nim_add_nim, Nat.cast_inj,
-        Nat.xor_left_inj]
-    · rwa [add_moveLeft_inr, moveLeft_nim', ne_eq, hk, grundyValue_nim_add_nim, Nat.cast_inj,
-        Nat.xor_right_inj]
-  -- For any `k < n ^^^ m`, either `nim (k ^^^ m) + nim m` or `nim n + nim (k ^^^ n)` is a left
-  -- option with Grundy value `k`.
+    apply leftMoves_add_cases i <;> intro j <;> have := (toLeftMovesNim_symm_lt j).ne
+    · simpa [grundyValue_nim_add_nim (toLeftMovesNim.symm j) y]
+    · simpa [grundyValue_nim_add_nim x (toLeftMovesNim.symm j)]
   · intro k hk
-    obtain ⟨k, rfl⟩ := Ordinal.lt_omega0.1 (hk.trans (Ordinal.nat_lt_omega0 _))
-    rw [Set.mem_Iio, Nat.cast_lt] at hk
-    obtain hk | hk := Nat.lt_xor_cases hk <;> rw [← natCast_lt] at hk
-    · use toLeftMovesAdd (Sum.inl (toLeftMovesNim ⟨_, hk⟩))
-      rw [Function.comp_apply, add_moveLeft_inl, moveLeft_nim', Equiv.symm_apply_apply,
-        grundyValue_nim_add_nim, Nat.xor_cancel_right]
-    · use toLeftMovesAdd (Sum.inr (toLeftMovesNim ⟨_, hk⟩))
-      rw [Function.comp_apply, add_moveLeft_inr, moveLeft_nim', Equiv.symm_apply_apply,
-        grundyValue_nim_add_nim, Nat.xor_comm, Nat.xor_cancel_right]
-termination_by (n, m)
-
-theorem nim_add_nim_equiv {n m : ℕ} : nim n + nim m ≈ nim (n ^^^ m) := by
+    obtain h | h := Nimber.lt_add_cases hk
+    · let a := toOrdinal (k + ∗y)
+      use toLeftMovesAdd (Sum.inl (toLeftMovesNim ⟨a, h⟩))
+      simp [a, grundyValue_nim_add_nim a y]
+    · let a := toOrdinal (k + ∗x)
+      use toLeftMovesAdd (Sum.inr (toLeftMovesNim ⟨a, h⟩))
+      simp [a, grundyValue_nim_add_nim x a, add_comm (∗x)]
+termination_by (x, y)
+
+theorem nim_add_nim_equiv (x y : Ordinal) :
+    nim x + nim y ≈ nim (toOrdinal (∗x + ∗y)) := by
   rw [← grundyValue_eq_iff_equiv_nim, grundyValue_nim_add_nim]
 
-theorem grundyValue_add (G H : PGame) [G.Impartial] [H.Impartial] {n m : ℕ} (hG : grundyValue G = n)
-    (hH : grundyValue H = m) : grundyValue (G + H) = n ^^^ m := by
-  rw [← nim_grundyValue (n ^^^ m), grundyValue_eq_iff_equiv]
-  refine Equiv.trans ?_ nim_add_nim_equiv
-  convert add_congr (equiv_nim_grundyValue G) (equiv_nim_grundyValue H) <;> simp only [hG, hH]
+@[simp]
+theorem grundyValue_add (G H : PGame) [G.Impartial] [H.Impartial] :
+    grundyValue (G + H) = grundyValue G + grundyValue H := by
+  rw [← (grundyValue G).toOrdinal_toNimber, ← (grundyValue H).toOrdinal_toNimber,
+    ← grundyValue_nim_add_nim, grundyValue_eq_iff_equiv]
+  exact add_congr (equiv_nim_grundyValue G) (equiv_nim_grundyValue H)
+
+theorem grundyValue_add_nat (G H : PGame) [G.Impartial] [H.Impartial] {n m : ℕ}
+    (hG : grundyValue G = ∗n) (hH : grundyValue H = ∗m) : grundyValue (G + H) = ∗(n ^^^ m) := by
+  rw [grundyValue_add, hG, hH, add_nat]
```

Comment:
This lemma is a bit weirdly stated, no? Do we need it?

Reply:
I don't think we actually need it, I just wanted to highlight that this previous version of the theorem is immediate from the one we have now.

Reply:
Yeah, I've removed it.